### PR TITLE
Split GitHub actions

### PR DESCRIPTION
**Pull request by igo95862**
Split Github actions by platform.

Only ubuntu workflow is enabled to run on any commit. (also on demand from any branch) Added my python binds build and mini run test. I also made it run with only local actions allowed by bypassing github checkout action. (unless @dankamongmen want to enable all actions)

Windows and MacOS workflows are on demand. (you can launch them from action page) They currently fail. I recommend keeping them that way until those platforms are supported.

### CODE REVIEW

**Comment by igo95862** [comment]
@michaelsbradleyjr

**Comment by michaelsbradleyjr** [comment]
> Split Github actions by platform.

Absent other considerations, splitting by platform seems to me to go against the grain of GitHub Actions (GHA). One of the beauties of the `matrix` facility of GHA, coupled with using a Unix-like build context for Windows (MSYS2) is that there can/should be (as a goal) only minor variance in the GHA steps of a workflow.

> Only ubuntu workflow is enabled to run on any commit. (also on demand from any branch)

I don't have strong feelings about that, but one thing to consider is that GHA doesn't provide an unlimited number of workflow minutes for open source projects.

By keeping it to PR branches and `master`, the idea is that contributors would run builds and tests locally; once a branch is working locally a contributor would create a PR.

Any one contributor may only build/test locally on 1/3 or 2/3 platforms because e.g. they don't have a mac or Windows machine. The workflow running for the PR then helps spotlight if the proposed changes will need adjustments re: the other platforms, maybe even prompting the contributor to ask for help in that regard. Of course, this effect is also achieved by running on every branch/commit.

Whether or not this project will bump up against the "free minutes limit" is, I guess, something that could be observed over time.

Details: https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#about-billing-for-github-actions

It's 2,000 minutes per month, and there are "minute multipliers" for macOS and Windows.

> Added my python binds build and mini run test.

Nice. I don't know much about the Python bindings, yours or the prior ones, but it's great to see active work in that area because of Python's popularity among TUI makers. 

> I also made it run with only local actions allowed by bypassing github checkout action. (unless @dankamongmen want to enable all actions)

It's obviously up to @dankamongmen. I personally consider [actions/checkout@v2](https://github.com/actions/checkout) (maintained by GitHub staff themselves) and [msys2/setup-msys2@v2](https://github.com/msys2/setup-msys2) to be quite safe, high quality, and very useful. 

But enabling 3rd party actions to be usable in GHA workflows does mean that folks merging PRs in this project should be mindful to check whether a PR changes a workflow, and if so whether it introduces an action that needs scrutiny. In general, any PR author adding an action or changing an existing action's configuration should justify why the change is necessary in the PR's description.

> Windows and MacOS workflows are on demand. (you can launch them from action page)

That's pretty cool!

> They currently fail.

I know, I spent hours writing them knowing they would fail, but making an effort for each one to get as far as possible without more involved changes in `CMakeLists.txt` and this project's source code.

> I recommend keeping them that way until those platforms are supported.

In the back-and-forth between myself and @dankamongmen in the comments of #1701, there was an idea that having an active workflow for macOS would provide a stepping stone toward that platform being supported by notcurses. But it's possible I misunderstood the exchange.

Consider that the workflow itself highlights what to install for macOS via `brew`, how to link `QR-Code-generator` (which is not available via `brew`), and then the failure of the `ctest` step gives some clue to what's going wrong or at least where to investigate.

I can also see merits of not activating the workflow until it works.

So I'll have to defer to @dankamongmen and others (inc. yourself) on this point.

File: .github/workflows/macos_test.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    env:
+      COLORTERM: truecolor
+      NPROC: 2
+      TERM: xterm
+    runs-on: macos-latest
+
+    steps:
+
+      - name: Install tools and libraries via Homebrew
+        if: matrix.platform.os == 'macos'
```
Comment (michaelsbradleyjr) [inline .github/workflows/macos_test.yml:18]: No matrix is defined in this workflow so the `if:` would bomb out, I think. But since each workflow in your PR is per platform, it could/should be removed anyway.

**Review by michaelsbradleyjr — COMMENTED** [review]
I've already left other general comments, but it's probably worth pointing out, again, that since 2/3 of the workflows in this PR use 3rd party actions, then a decision should probably be made as to whether that's okay in this repository or not.

On the other hand, I guess could the workflows could be left as is and folks who fork the repo can choose to enable 3rd party actions in their forks, i.e. to be able to run the workflows for macOS and Windows.

File: .github/workflows/macos_test.yml (hunk 1 above)
Comment (igo95862) [reply .github/workflows/macos_test.yml]: Should be fixed now.

**Comment by igo95862** [comment]
> only minor variance in the GHA steps of a workflow.

There are already giant variations but inside the nested `if` statements which made it difficult to read and modify.  

> Of course, this effect is also achieved by running on every branch/commit.

Old workflow also ran every commit to master and pull request. I only added the manual launch option which only the owner of repository can launch. (I found it useful in other projects as it can be run from any branch)

> active workflow for macOS

BTW I think if you modify the MacOS workflow in the pull request for MacOS the modified action will run inside the Pull Request. If you look this Pull Request ran its own new action.

Another advantage of separated workflows is that the `on:` key can  be specified per platform. For example, since MacOS drains 10x minutes make it run every couple days instead of commits to master. (you can probably achieve that with more `if` hacks but that would be really messy)

**Comment by michaelsbradleyjr** [comment]
> > only minor variance in the GHA steps of a workflow.
> 
> There are already giant variations but inside the nested `if` statements which made it difficult to read and modify.

True, this is an eye sore: https://github.com/dankamongmen/notcurses/blob/master/.github/workflows/tests.yml#L99-L120.

It could be / should have been broken out into its own step named `QR-Code-generator`, or into two steps with postfixes `macos` and `windows`.

The  proper `cmake` part below those lines has some branching too, though I hoped in time to have a unified invocation by way of logic in `CMakeLists.txt`. I just haven't spent much time with `cmake` yet, so it was easier for me to accomplish on a first pass in the embedded Bash script.
 
> Another advantage of separated workflows is that the `on:` key can be specified per platform. For example, since MacOS drains 10x minutes make it run every couple days instead of commits to master. (you can probably achieve that with more `if` hacks but that would be really messy)

Sure, that is indeed a nice plus in favor of separating the workflows by platform.

Whichever way folks feel is the best for this project is how it should roll, and I'm just happy to helped get the ball rolling and look forward to contributing in the future as I'm able.

**Comment by dankamongmen** [comment]
look, y'all, i'm a pretty decent c coder when allowed to remove myself to a cave and eat eels and not talk to people for a few days, but i am clearly far far behind the both of you in this subject matter, and i don't think i'm going to change that in the next fifteen minutes.

you're both collaborators and have commit rights, so if you're waiting for me, don't.

i took a brief glance over this, and mainly just came away happy that skilled people are putting their valuable time into my much-beloved project. merge at will, gentlemen.

**Comment by dankamongmen** [comment]
oh you've actually got @michaelsbradleyjr as a reviewer, and not me, and he has not yet approved. ok, the system appears to be working! i'm going to go eat some fish.

**Comment by dankamongmen** [comment]
i *am* interested in how/if this will integrate with/complement/replace the existing CI at https://drone.dsscaw.com:4443/dankamongmen/notcurses, which has treated me well thus far, but to which i am in no way wedded (though i'll need keep running the server for my other projects). you are of course welcome to press these computing resources (my personal servers) into service.

**Comment by michaelsbradleyjr** [comment]
@dankamongmen 
> oh you've actually got @michaelsbradleyjr as a reviewer, and not me, and he has not yet approved. ok, the system appears to be working! i'm going to go eat some fish.

A key question, among others pertinent to this PR (and the PR of mine you previously merged) is...

If it's okay for 3rd party GitHub Actions (GHA) actions to be allowed in this repository GHA's workflow/s? That's going to require an admin-level decision and checkbox-ing in the admin-level settings for this repo.

Honesty, I think it's okay, almost zero risk, since future unwanted/unvetted GHA are very unlikely to be activated by PR merges of 3rd party PRs, unless official contibutors in this project are just terribly sloppy in reviewing and merging PRs.

As for splitting the GHA workflow (I submitted and you merged into `master`) into three workflows per this PR...

Well, I'll simply say I'm not 100% convinced of the merits of that approach, while also agreeing the workflow I originally submitted could use some TLC.

And that also there are objective merits of splitting the workflows by platform according to what @igo95862 has done and commented upon in this PR.

In summary, I'd suggest allowing for another day's round of feedback, and then let's go with whatever strikes your (@dankamongmen's) gut as the best approach in the near term.

**Comment by michaelsbradleyjr** [comment]
Addendum: let's try to boil it down into minimalist pros and cons.

**Comment by michaelsbradleyjr** [comment]
I gave it some more thought today (a day late, sorry), let's go with the "split actions" @igo95862 has cooked up (am approving this PR). If it ends up having more downsides than upsides then I guess we could always go back to a unified action like the one I authored originally.

The only thing is that if the macOS and Windows actions are ever to run in this repo, then permission for 3rd party actions needs to be activated. Maybe we can get a decision on that @dankamongmen?

If 3rd party actions are a no-go from repo admin's / admins' perspective/s (not sure if there's more than one admin-level contributor), then the macOS workflow could be adjusted to manually invoke `git`, as done in this PR's workflow for Linux.

However, for Windows there's a bit more involved as the MSYS2 installer would need to be downloaded and run in headless mode, `pacman -S ...` would need to be manually invoked, etc. And then there's the question of trust being placed in the MSYS2 installer... and the people who wrote that installer also maintain the [msys2/setup-msys2@v2](https://github.com/msys2/setup-msys2) action... so what's the point?

Also, I think it's known to parties involved in this PR discussion, but the macOS and Windows workflows won't run unless manually triggered by someone who is a contributor (or maybe that requires admin-level, not sure). I believe that's something that people who (soft) fork this repo can relax in their fork's GitHub settings, but not sure yet, will double-check on my fork if/when this PR lands in `master`.

**Review by michaelsbradleyjr — APPROVED** [review]
Let's do it!
